File: aplicaciones_demostrativas/mw/CS/CSmwEIDTest_VisualStudio-2010/CSmwEIDTest/OcspClient.cs

using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Net;
using Org.BouncyCastle.Asn1;
using Org.BouncyCastle.Asn1.Ocsp;
using Org.BouncyCastle.Asn1.X509;
using Org.BouncyCastle.Math;
using Org.BouncyCastle.Ocsp;
using Org.BouncyCastle.X509;


namespace CSmwEIDTest
{

public enum CertificateStatus { Good = 0, Revoked = 1, Unknown = 2 };

    class X509TestPublicKeyCertificate : X509Certificate
    {
        public X509TestPublicKeyCertificate(X509CertificateStructure in_Cert)
            : base(in_Cert)
        {
        }
    }

    class OcspClient
    {

        public readonly int BufferSize = 4096 * 8;
        private readonly int MaxClockSkew = 36000000;

        private X509TestPublicKeyCertificate m_PublicKeyCertificate = null;
        public X509TestPublicKeyCertificate PublicKeyCertificate
        {
            get
            {
                return m_PublicKeyCertificate;
            }
        }

        public X509Certificate LeerCertificado(String in_Filename)
        {
            X509CertificateParser certParser = new X509CertificateParser();

            Stream stream = new FileStream(in_Filename, FileMode.Open);
            X509Certificate certificado = certParser.ReadCertificate(stream);
            stream.Close();

            return certificado;
        }

        public OcspClient(byte[] in_CertificateInput)
        {
            X509CertificateParser certParser = new X509CertificateParser();
            X509Certificate certificado = (X509Certificate)certParser.ReadCertificate(in_CertificateInput);
            m_PublicKeyCertificate = new X509TestPublicKeyCertificate(certificado.CertificateStructure);

        }

        public CertificateStatus ConsultarEstadoDeCertificado(X509Certificate in_Certificado, X509Certificate in_CertificadoEmisor)
        {
            List<string> urls = GetAuthorityInformationAccessOcspUrl(in_Certificado);

            if (urls.Count == 0)
            {
                throw new Exception("No se encontro ningun OCSP url en el certificado.");
            }

            string url = urls[0];

            Console.WriteLine("Consultando '" + url + "'...");

            OcspReq req = GenerarRequestOCSP(in_CertificadoEmisor, in_Certificado.SerialNumber);

            byte[] binaryResp = PostData(url, req.GetEncoded(), "application/ocsp-request", "application/ocsp-response");

            return ProcesarRespuestaOcsp(in_Certificado, in_CertificadoEmisor, binaryResp);
        }



        public byte[] PostData(string in_Url, byte[] in_Data, string in_ContentType, string in_Accept)
        {
            HttpWebRequest request = (HttpWebRequest)WebRequest.Create(in_Url);
            request.Method = "POST";
            request.ContentType = in_ContentType;
            request.ContentLength = in_Data.Length;
            request.Accept = in_Accept;
            Stream stream = request.GetRequestStream();
            stream.Write(in_Data, 0, in_Data.Length);
            stream.Close();
            HttpWebResponse response = (HttpWebResponse)request.GetResponse();
            Stream respStream = response.GetResponseStream();
            byte[] resp = ToByteArray(respStream);
            respStream.Close();

            return resp;
        }

        public byte[] ToByteArray(Stream in_Stream)
        {
            byte[] buffer = new byte[BufferSize];
            MemoryStream ms = new MemoryStream();

            int read = 0;

            while ((read = in_Stream.Read(buffer, 0, buffer.Length)) > 0)
            {
                ms.Write(buffer, 0, read);
            }

            return ms.ToArray();
        }


        public static List<string> GetAuthorityInformationAccessOcspUrl(X509Certificate in_Certificado)
        {
            List<string> ocspUrls = new List<string>();

            try
            {
                Asn1Object obj = ObtenerValorDeExtension(in_Certificado, X509Extensions.AuthorityInfoAccess.Id);

                if (obj == null)
                {
                    return null;
                }


                Asn1Sequence s = (Asn1Sequence)obj;
                IEnumerator elementos = s.GetEnumerator();

                while (elementos.MoveNext())
                {
                    Asn1Sequence elemento = (Asn1Sequence)elementos.Current;
                    DerObjectIdentifier oid = (DerObjectIdentifier)elemento[0];

                    if (oid.Id.Equals("1.3.6.1.5.5.7.48.1")) // Ocsp?
                    {
                        Asn1TaggedObject objetoTagged = (Asn1TaggedObject)elemento[1];
                        GeneralName gn = (GeneralName)GeneralName.GetInstance(objetoTagged);
                        ocspUrls.Add(((DerIA5String)DerIA5String.GetInstance(gn.Name)).GetString());
                    }
                }
            }
            catch (Exception e)
            {
                throw new Exception("Error en AuthorityInformationAccess.", e);
            }

            return ocspUrls;
        }

        protected static Asn1Object ObtenerValorDeExtension(X509Certificate in_Certificado,
                string oid)
        {
            if (in_Certificado == null)
            {
                return null;
            }

            byte[] bytes = in_Certificado.GetExtensionValue(new DerObjectIdentifier(oid)).GetOctets();

            if (bytes == null)
            {
                return null;
            }

            Asn1InputStream aIn = new Asn1InputStream(bytes);

            return aIn.ReadObject();
        }


        private CertificateStatus ProcesarRespuestaOcsp(X509Certificate in_Certificado, X509Certificate in_CertificadoEmisor, byte[] in_BytesRespuesta)
        {
            OcspResp r = new OcspResp(in_BytesRespuesta);
            CertificateStatus estado = CertificateStatus.Unknown;

            switch (r.Status)
            {
                case OcspRespStatus.Successful:
                    BasicOcspResp or = (BasicOcspResp)r.GetResponseObject();


                    if (or.Responses.Length == 1)
                    {
                        SingleResp resp = or.Responses[0];

                        ValidarCertificateId(in_CertificadoEmisor, in_Certificado, resp.GetCertID());
                        Object estadoCertificado = resp.GetCertStatus();

                        if (estadoCertificado == Org.BouncyCastle.Ocsp.CertificateStatus.Good)
                        {
                            estado = CertificateStatus.Good;
                        }
                        else if (estadoCertificado is Org.BouncyCastle.Ocsp.RevokedStatus)
                        {
                            estado = CertificateStatus.Revoked;
                        }
                        else if (estadoCertificado is Org.BouncyCastle.Ocsp.UnknownStatus)
                        {
                            estado = CertificateStatus.Unknown;
                        }
                    }
                    break;
                default:
                    throw new Exception("Status desconocido'" + r.Status + "'.");
            }

            return estado;
        }

        private void ValidateResponse(BasicOcspResp in_OcspResp, X509Certificate in_CertificadoEmisor)
        {
            ValidarResponseSignature(in_OcspResp, in_CertificadoEmisor.GetPublicKey());
            ValidarSignerAuthorization(in_CertificadoEmisor, in_OcspResp.GetCerts()[0]);
        }


        private void ValidarSignerAuthorization(X509Certificate in_CertificadoEmisor, X509Certificate in_CertificadoFirmante)
        {

            if (!(in_CertificadoEmisor.IssuerDN.Equivalent(in_CertificadoFirmante.IssuerDN) && in_CertificadoEmisor.SerialNumber.Equals(in_CertificadoFirmante.SerialNumber)))
            {
                throw new Exception("OCSP signer Invalido");
            }
        }

        private void ValidarResponseSignature(BasicOcspResp in_OcspResp, Org.BouncyCastle.Crypto.AsymmetricKeyParameter in_PublicKey)
        {
            if (!in_OcspResp.Verify(in_PublicKey))
            {
                throw new Exception("Firma OCSP Invalida");
            }
        }


        private void ValidarCertificateId(X509Certificate in_CertificadoEmisor, X509Certificate in_Certificado, CertificateID in_IDCertificado)
        {
            CertificateID idEsperado = new CertificateID(CertificateID.HashSha1, in_CertificadoEmisor, in_Certificado.SerialNumber);

            if (!idEsperado.SerialNumber.Equals(in_IDCertificado.SerialNumber))
            {
                throw new Exception("ID de Certificado invalido");
            }

            if (!Org.BouncyCastle.Utilities.Arrays.AreEqual(idEsperado.GetIssuerNameHash(), in_IDCertificado.GetIssuerNameHash()))
            {
                throw new Exception("Certificado Emisor invalido");
            }

        }

        private OcspReq GenerarRequestOCSP(X509Certificate in_CertificadoEmisor, BigInteger in_NumeroSerie)
        {
            CertificateID id = new CertificateID(CertificateID.HashSha1, in_CertificadoEmisor, in_NumeroSerie);
            return GenerarRequestOCSP(id);
        }

        private OcspReq GenerarRequestOCSP(CertificateID in_Id)
        {
            OcspReqGenerator ocspRequestGenerador = new OcspReqGenerator();

            ocspRequestGenerador.AddRequest(in_Id);

            BigInteger nonce = BigInteger.ValueOf(new DateTime().Ticks);

            ArrayList oids = new ArrayList();
            Hashtable valores = new Hashtable();

            oids.Add(OcspObjectIdentifiers.PkixOcsp);

            Asn1OctetString asn1 = new DerOctetString(new DerOctetString(new byte[] { 1, 3, 6, 1, 5, 5, 7, 48, 1, 1 }));

            valores.Add(OcspObjectIdentifiers.PkixOcsp, new X509Extension(false, asn1));
            ocspRequestGenerador.SetRequestExtensions(new X509Extensions(oids, valores));

            return ocspRequestGenerador.Generate();
        }
    }
} 

File: aplicaciones_demostrativas/trust-service/AplicacionDemoApplet/AplicacionDemoApplet-ejb/src/main/java/com/mycompany/OcspClient.java
package com.mycompany;

import org.bouncycastle.asn1.*;
import org.bouncycastle.asn1.ocsp.OCSPObjectIdentifiers;
import org.bouncycastle.asn1.x509.*;
import org.bouncycastle.ocsp.*;

import java.io.*;
import java.math.BigInteger;
import java.net.HttpURLConnection;
import java.net.URL;
import java.security.Security;
import java.security.cert.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Vector;


public class OcspClient {

    private X509Certificate m_PublicKeyCertificate;

    public X509Certificate getPublicKeyCertificate() {
        return m_PublicKeyCertificate;
    }

    public OcspClient(byte[] in_CertificateInput) {
        try {
            CertificateFactory factory = CertificateFactory.getInstance("X.509");
            ByteArrayInputStream certificateInputStream = new ByteArrayInputStream(in_CertificateInput);

            m_PublicKeyCertificate = (X509Certificate) factory.generateCertificate(certificateInputStream);

        } catch (CertificateException ex) {
            System.out.println(ex);
        }
    }

    public OcspClient(X509Certificate  in_CertificateInput) {
        try {
            m_PublicKeyCertificate = in_CertificateInput;
        } catch (Exception ex) {
            System.out.println(ex);
        }
    }

        public X509Certificate leerCertificado(String filename)
        throws Exception {
        CertificateFactory factory = CertificateFactory.getInstance("X.509");

        FileInputStream stream = new FileInputStream(filename);
        X509Certificate certificado = (X509Certificate) factory.generateCertificate(stream);
        stream.close();

        return certificado;
    }

    public CertificateStatus consultarEstadoDeCertificado(X509Certificate in_Certificado, X509Certificate in_CertificadoEmisor)
            throws Exception {

        OCSPReq request = generarRequestOCSP(in_CertificadoEmisor, in_Certificado.getSerialNumber());

        List<String> locations = GetAuthorityInformationAccessOcspUrl(in_Certificado);

        for (String serviceUrl : locations) {

            SingleResp[] responses;
            try {
                OCSPResp ocspResponse = obtenerRespuestaOCSP(serviceUrl, request);
                if (OCSPRespStatus.SUCCESSFUL != ocspResponse.getStatus()) {
                    continue;
                }

                BasicOCSPResp basicResponse = (BasicOCSPResp) ocspResponse.getResponseObject();
                responses = (basicResponse == null) ? null : basicResponse.getResponses();

            } catch (Exception e) {
                continue;
            }

            if (responses != null && responses.length == 1) {
                SingleResp resp = responses[0];
                CertificateStatus status = obtenerEstado(resp);
                return status;
            }
        }
        throw new Exception("No se puede obtener el Estado de Revocacion del Servidor OCSP.");
    }

    private CertificateStatus obtenerEstado(SingleResp resp) throws Exception {
        Object status = resp.getCertStatus();
        if (status == null) {
            return CertificateStatus.Good;
        } else if (status instanceof org.bouncycastle.ocsp.RevokedStatus) {
            return CertificateStatus.Revoked;
        } else if (status instanceof org.bouncycastle.ocsp.UnknownStatus) {
            return CertificateStatus.UNKNOWN;
        }
        return CertificateStatus.UNKNOWN;
    }

    private OCSPResp obtenerRespuestaOCSP(String ocspUrl,
            OCSPReq request) throws Exception {
        try {

            byte[] array = request.getEncoded();
            if (ocspUrl.startsWith("http")) {
                HttpURLConnection con;
                URL url = new URL(ocspUrl);
                con = (HttpURLConnection) url.openConnection();
                con.setRequestProperty("Content-Type", "application/ocsp-request");
                con.setRequestProperty("Accept", "application/ocsp-response");
                con.setDoOutput(true);
                OutputStream out = con.getOutputStream();
                DataOutputStream dataOut = new DataOutputStream(new BufferedOutputStream(out));
                dataOut.write(array);

                dataOut.flush();
                dataOut.close();


                if (con.getResponseCode() / 100 != 2) {
                    throw new Exception("Error en la respuesta OCSP."
                            + "Codigo de Respuesta:  " + con.getResponseCode());
                }


                InputStream in = (InputStream) con.getContent();
                return new OCSPResp(in);
            } else {
                throw new Exception("Solamente http es soportado para OCSP");
            }
        } catch (IOException e) {
            throw new Exception("No se puede obtener  respuesta OCSP de: " + ocspUrl, e);
        }
    }


    private OCSPReq generarRequestOCSP(X509Certificate in_CertificadoEmisor, BigInteger in_NumeroSerie)
            throws Exception {

        Security.addProvider(new org.bouncycastle.jce.provider.BouncyCastleProvider());
        try {

            CertificateID id = new CertificateID(CertificateID.HASH_SHA1, in_CertificadoEmisor, in_NumeroSerie);


            OCSPReqGenerator ocspRequestGenerator = new OCSPReqGenerator();
            ocspRequestGenerator.addRequest(id);

            //BigInteger nonce = BigInteger.valueOf(System.currentTimeMillis());
            Vector<DERObjectIdentifier> objectIdentifiers = new Vector<DERObjectIdentifier>();
            Vector<org.bouncycastle.asn1.x509.X509Extension> values = new Vector<org.bouncycastle.asn1.x509.X509Extension>();

            ASN1OctetString asn1 = new DEROctetString(new DEROctetString(new byte[] { 1, 3, 6, 1, 5, 5, 7, 48, 1, 1 }));

            //objectIdentifiers.add(OCSPObjectIdentifiers.id_pkix_ocsp_nonce);
            //values.add(new org.bouncycastle.asn1.x509.X509Extension(false, new DEROctetString(nonce.toByteArray())));
            objectIdentifiers.add(OCSPObjectIdentifiers.id_pkix_ocsp);
            values.add(new org.bouncycastle.asn1.x509.X509Extension(false, asn1));
            ocspRequestGenerator.setRequestExtensions(new X509Extensions(objectIdentifiers, values));

            return ocspRequestGenerator.generate();
        } catch (OCSPException e) {
            throw new Exception("No se puede generar OCSP Request con el "
                    + "certificado", e);
        }
    }


    private List<String> GetAuthorityInformationAccessOcspUrl(X509Certificate cert) throws Exception {


        byte[] valorExtensionAIA = cert.getExtensionValue(X509Extensions.AuthorityInfoAccess.getId());
        if (valorExtensionAIA == null) {
            throw new Exception("El Certificado no contiene AuthorityInformationAccess");
        }

        ASN1InputStream asn1In = new ASN1InputStream(valorExtensionAIA);
        AuthorityInformationAccess authorityInformationAccess;

        try {
            DEROctetString aiaDEROctetString = (DEROctetString) (asn1In.readObject());
            ASN1InputStream asn1InOctets = new ASN1InputStream(aiaDEROctetString.getOctets());
            ASN1Sequence aiaASN1Sequence = (ASN1Sequence) asn1InOctets.readObject();
            authorityInformationAccess = AuthorityInformationAccess.getInstance(aiaASN1Sequence);
        } catch (IOException e) {
            throw new Exception("No se puede leer el certificado para obtener OCSP URLs", e);
        }

        List<String> ocspUrlList = new ArrayList<String>();
        AccessDescription[] accessDescriptions = authorityInformationAccess.getAccessDescriptions();
        for (AccessDescription accessDescription : accessDescriptions) {

            String oid = accessDescription.getAccessMethod().getId();
            GeneralName gn = accessDescription.getAccessLocation();
            if (gn.getTagNo() == GeneralName.uniformResourceIdentifier &&
                    oid.equals("1.3.6.1.5.5.7.48.1")) {  //ocsp?
                DERIA5String str = DERIA5String.getInstance(gn.getName());
                String accessLocation = str.getString();
                ocspUrlList.add(accessLocation);
            }
        }
        if (ocspUrlList.isEmpty()) {
            throw new Exception("No se puede obtener OCSP urls del certificado");
        }

        return ocspUrlList;
    }
}
